File: lib/lambda/rules/aws_cis_3_14.ts
import { cloudTrailRecord, uguisuRule, detection } from "../models";

// https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#cis-3.14-remediation

export class rule extends uguisuRule {
  constructor() {
    super({
      id: "aws_cis_3.14",
      title: "VPC changes",
      description:
        "AWS CIS benchmark 3.14 recommend to ensure a log metric filter and alarm exist for VPC changes",
      severity: "medium",
    });
  }

  detect(record: cloudTrailRecord): detection | null {
    const targetEvents: { [key: string]: boolean } = {
      CreateVpc: true,
      DeleteVpc: true,
      ModifyVpcAttribute: true,
      AcceptVpcPeeringConnection: true,
      CreateVpcPeeringConnection: true,
      DeleteVpcPeeringConnection: true,
      RejectVpcPeeringConnection: true,
      AttachClassicLinkVpc: true,
      DetachClassicLinkVpc: true,
      DisableVpcClassicLink: true,
      EnableVpcClassicLink: true,
    };

    if (targetEvents[record.eventName] === true) {
      return {
        rule: this,
        event: record,
      };
    }

    return null;
  }
}

File: pkg/rules/aws_cis_3_14.go
package rules

import (
	"github.com/m-mizutani/uguisu/pkg/models"
)

// https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#cis-3.14-remediation
type awsCIS3_14 struct {
	targetEvents map[string]bool
}

func newAwsCIS3_14() models.Rule {
	return &awsCIS3_14{
		targetEvents: map[string]bool{
			"CreateVpc":                  true,
			"DeleteVpc":                  true,
			"ModifyVpcAttribute":         true,
			"AcceptVpcPeeringConnection": true,
			"CreateVpcPeeringConnection": true,
			"DeleteVpcPeeringConnection": true,
			"RejectVpcPeeringConnection": true,
			"AttachClassicLinkVpc":       true,
			"DetachClassicLinkVpc":       true,
			"DisableVpcClassicLink":      true,
			"EnableVpcClassicLink":       true},
	}
}

func (x *awsCIS3_14) ID() string                { return "aws_cis_3.14" }
func (x *awsCIS3_14) Title() string             { return "Route table changes" }
func (x *awsCIS3_14) Severity() models.Severity { return models.SeverityMedium }
func (x *awsCIS3_14) Description() string {
	return "AWS CIS benchmark 3.14 recommend to ensure a log metric filter and alarm exist for VPC changes"
}
func (x *awsCIS3_14) Match(record *models.CloudTrailRecord) bool {
	return x.targetEvents[record.EventName]
}
